go/batch-submitter/drivers/sequencer/encoding.go

```diff
@@ -168,6 +239,34 @@ func (p *AppendSequencerBatchParams) Read(r io.Reader) error {
 		p.Contexts = append(p.Contexts, batchContext)
 	}
 
+	p.Type = BatchTypeLegacy
+
+	// Handle backwards compatible batch types
+	if len(p.Contexts) > 0 && p.Contexts[0].Timestamp == 0 {
+		switch p.Contexts[0].BlockNumber {
+		case 0:
+			// zlib compressed transaction data
+			p.Type = BatchTypeZlib
+			// remove the first dummy context
+			p.Contexts = p.Contexts[1:]
+			numContexts--
+
+			zr, err := zlib.NewReader(r)
+			if err == io.EOF || err == io.ErrUnexpectedEOF {
+				return nil
```

Comment:
Are `io.EOF` and `io.ErrUnexpectedEOF` errors that we expect to see before reading any data from the batch context?

Reply:
The reason that there is an `io.ErrUnexpectedEOF` here is because the existing unit tests include malformed batches, ie ones with no contexts but with transactions:
https://github.com/ethereum-optimism/optimism/blob/43f16dd06b7173a6b8a6703fec984df8b2ea91c2/go/batch-submitter/drivers/sequencer/encoding_test.go#L88

For the txs to be able to execute, there must be a context. Perhaps we should add some validation that the shape of the batch is correct and instead return an error when a malformed batch is being deserialized

Reply:
Yeah, I think that adding validation is a better choice here.

Reply:
I've added some batch validation with test coverage